src/stats.ml

```diff
@@ -1,53 +1,119 @@
+    (List.map (fun { Rrd_timescales.num_intervals; interval_in_steps } ->
+      if interval_in_steps > 1 && use_min_max then [
+        Rrd.rra_create Rrd.CF_Average num_intervals interval_in_steps 1.0;
+        Rrd.rra_create Rrd.CF_Min num_intervals interval_in_steps 1.0;
+        Rrd.rra_create Rrd.CF_Max num_intervals interval_in_steps 1.0;
+      ] else [Rrd.rra_create Rrd.CF_Average num_intervals interval_in_steps 0.5]
+    ) timescales)
+  )
+
+let step = 5
+
+module Ds = struct
+  type t = {
+    name: string;
+    description: string;
+    value: Rrd.ds_value_type;
+    ty: Rrd.ds_type;
+    max: float;
+    min: float;
+    units: string;
+  }
+  let make ~name ~description ~value ~ty ~units
+    ?(min = neg_infinity) ?(max = infinity) () = {
+    name; description; value; ty; min; max; units
+  }
+end
+
+let make_dss stats = [
+  Ds.make ~name:"free_words" ~units:"words"
+    ~description:"Number of words in the free list"
+    ~value:(Rrd.VT_Int64 (Int64.of_int stats.Gc.free_words)) ~ty:Rrd.Gauge ~min:0.0 ();
+  Ds.make ~name:"live_words" ~units:"words"
+    ~description:"Number of words of live data in the major heap, including the header words."
+    ~value:(Rrd.VT_Int64 (Int64.of_int stats.Gc.live_words)) ~ty:Rrd.Gauge ~min:0.0 ();
+  ]
+
+(** Create a rrd *)
+let create_fresh_rrd use_min_max dss =
+  let rras = create_rras use_min_max in
+  let dss = Array.of_list (List.map (fun ds ->
+      Rrd.ds_create ds.Ds.name ds.Ds.ty ~mrhb:300.0 ~max:ds.Ds.max
+      ~min:ds.Ds.min Rrd.VT_Unknown
+    ) dss) in
+  let rrd = Rrd.rrd_create dss rras (Int64.of_int step) (Clock.time ()) in
+  rrd
+
+let update_rrds timestamp dss rrd =
+  Rrd.ds_update_named rrd timestamp ~new_domid:false
+    (List.map (fun ds -> ds.Ds.name, (ds.Ds.value, fun x -> x)) dss)
+
+let rrd = create_fresh_rrd true (make_dss (Gc.stat ()))
 
 let start ~sleep =
-  let gather () =
-    let stat = Gc.stat () in
-    if Queue.length stats >= history then ignore (Queue.pop stats);
-    Queue.push stat stats
-  in
   let rec loop () =
-    gather ();
-    sleep delay >>= fun () ->
-    loop ()
-  in
+    let timestamp = Clock.time () in
+    update_rrds timestamp (make_dss (Gc.stat ())) rrd;
+    sleep 5. >>= fun () ->
+    loop () in
   Lwt.async loop
 
 let page () =
-  let stats = Queue.fold (fun acc s -> s :: acc) [] stats in
-  html_of_stats stats
+  let timescales = List.map (fun t ->
+    let uri = "?timescale=" ^ t.Rrd_timescales.name in
+    <:html<
+    <a href="$str:uri$">$str:t.Rrd_timescales.name$</a>
+    >>
+  ) timescales in
+  <:html<
+    <head>
+      <meta charset="utf-8" />
+      <link href="https://cdnjs.cloudflare.com/ajax/libs/c3/0.4.10/c3.css" rel="stylesheet" type="text/css"/>
+      <script src="https://cdnjs.cloudflare.com/ajax/libs/d3/3.5.5/d3.min.js" charset="utf-8"></script>
+      <script src="https://cdnjs.cloudflare.com/ajax/libs/c3/0.4.10/c3.min.js"></script>
+      <script src="/js/stats/main.js"> </script>
```

Comment:
Note the external references here. I could add C3 and D3 to this repo if you'd prefer.